### PR TITLE
Remaining Doctype Errors

### PR DESCRIPTION
**Pull request by diervo**
@inikulin Let me know if the names make sense to you.

### CODE REVIEW

File: lib/tokenizer/index.js [hunk 1]
```diff
@@ -1926,28 +1926,33 @@ _[AFTER_DOCTYPE_PUBLIC_KEYWORD_STATE] = function afterDoctypePublicKeywordState(
         this.state = BEFORE_DOCTYPE_PUBLIC_IDENTIFIER_STATE;
 
     else if (cp === $.QUOTATION_MARK) {
+        this._err(ERR.missingSeparationCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: It's better be named `missing-whitespace-after-doctype-public-keyword`

File: lib/tokenizer/index.js [hunk 2]
```diff
@@ -1926,28 +1926,33 @@ _[AFTER_DOCTYPE_PUBLIC_KEYWORD_STATE] = function afterDoctypePublicKeywordState(
...
         this.currentToken.publicId = '';
         this.state = DOCTYPE_PUBLIC_IDENTIFIER_DOUBLE_QUOTED_STATE;
     }
 
     else if (cp === $.APOSTROPHE) {
+        this._err(ERR.missingSeparationCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: Same as above

File: lib/tokenizer/index.js [hunk 3]
```diff
@@ -1926,28 +1926,33 @@ _[AFTER_DOCTYPE_PUBLIC_KEYWORD_STATE] = function afterDoctypePublicKeywordState(
...
         this.currentToken.publicId = '';
         this.state = DOCTYPE_PUBLIC_IDENTIFIER_SINGLE_QUOTED_STATE;
     }
 
     else if (cp === $.GREATER_THAN_SIGN) {
+        this._err(ERR.unexpectedClosingOfDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: We've used term `abrupt` previously for such cases, so it's an `abrupt-closing-of-doctype`.

File: lib/tokenizer/index.js [hunk 4]
```diff
@@ -1926,28 +1926,33 @@ _[AFTER_DOCTYPE_PUBLIC_KEYWORD_STATE] = function afterDoctypePublicKeywordState(
...
 
     else if (cp === $.EOF) {
+        this._err(ERR.eofInDoctype);
         this.currentToken.forceQuirks = true;
         this._emitCurrentToken();
         this._emitEOFToken();
     }
 
     else {
+        this._err(ERR.unexpectedCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: It's `missing-quote-around-doctype-public-identifier`

File: lib/tokenizer/index.js [hunk 5]
```diff
@@ -1971,18 +1976,21 @@ _[BEFORE_DOCTYPE_PUBLIC_IDENTIFIER_STATE] = function beforeDoctypePublicIdentifi
     }
 
     else if (cp === $.GREATER_THAN_SIGN) {
+        this._err(ERR.unexpectedClosingOfDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `abrupt-closing-of-doctype`

File: lib/tokenizer/index.js [hunk 6]
```diff
@@ -1971,18 +1976,21 @@ _[BEFORE_DOCTYPE_PUBLIC_IDENTIFIER_STATE] = function beforeDoctypePublicIdentifi
...
 
     else if (cp === $.EOF) {
+        this._err(ERR.eofInDoctype);
         this.currentToken.forceQuirks = true;
         this._emitCurrentToken();
         this._emitEOFToken();
     }
 
     else {
+        this._err(ERR.unexpectedCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `missing-quote-before-doctype-public-identifier`

File: lib/tokenizer/index.js [hunk 7]
```diff
@@ -1995,16 +2003,20 @@ _[DOCTYPE_PUBLIC_IDENTIFIER_DOUBLE_QUOTED_STATE] = function doctypePublicIdentif
...
         this.state = AFTER_DOCTYPE_PUBLIC_IDENTIFIER_STATE;
 
-    else if (cp === $.NULL)
+    else if (cp === $.NULL) {
+        this._err(ERR.unexpectedNullCharacter);
         this.currentToken.publicId += UNICODE.REPLACEMENT_CHARACTER;
+    }
 
     else if (cp === $.GREATER_THAN_SIGN) {
+        this._err(ERR.unexpectedClosingOfDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `abrupt-closing-of-doctype`

File: lib/tokenizer/index.js [hunk 8]
```diff
@@ -2021,16 +2033,20 @@ _[DOCTYPE_PUBLIC_IDENTIFIER_SINGLE_QUOTED_STATE] = function doctypePublicIdentif
...
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `abrupt-closing-of-doctype`

File: lib/tokenizer/index.js [hunk 9]
```diff
@@ -2052,28 +2068,31 @@ _[AFTER_DOCTYPE_PUBLIC_IDENTIFIER_STATE] = function afterDoctypePublicIdentifier
     }
 
     else if (cp === $.QUOTATION_MARK) {
+        this._err(ERR.missingSeparationCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `missing-whitespace-after-doctype-public-identifier`

File: lib/tokenizer/index.js [hunk 10]
```diff
@@ -2052,28 +2068,31 @@ _[AFTER_DOCTYPE_PUBLIC_IDENTIFIER_STATE] = function afterDoctypePublicIdentifier
...
         this.currentToken.systemId = '';
         this.state = DOCTYPE_SYSTEM_IDENTIFIER_DOUBLE_QUOTED_STATE;
     }
 
     else if (cp === $.APOSTROPHE) {
+        this._err(ERR.missingSeparationCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `missing-whitespace-between-doctype-public-and-system-identifiers`

File: lib/tokenizer/index.js [hunk 11]
```diff
@@ -2052,28 +2068,31 @@ _[AFTER_DOCTYPE_PUBLIC_IDENTIFIER_STATE] = function afterDoctypePublicIdentifier
...
 
     else if (cp === $.EOF) {
+        this._err(ERR.eofInDoctype);
         this.currentToken.forceQuirks = true;
         this._emitCurrentToken();
         this._emitEOFToken();
     }
 
     else {
+        this._err(ERR.unexpectedCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `missing-quote-before-doctype-system-identifier`

File: lib/tokenizer/index.js [hunk 12]
```diff
@@ -2096,7 +2115,15 @@ _[BETWEEN_DOCTYPE_PUBLIC_AND_SYSTEM_IDENTIFIERS_STATE] = function betweenDoctype
...
 
+    else if (cp === $.EOF) {
+        this._err(ERR.eofInDoctype);
+        this.currentToken.forceQuirks = true;
+        this._emitCurrentToken();
+        this._emitEOFToken();
+    }
+
     else {
+        this._err(ERR.unexpectedCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `missing-quote-before-doctype-system-identifier`

File: lib/tokenizer/index.js [hunk 13]
```diff
@@ -2110,28 +2137,33 @@ _[AFTER_DOCTYPE_SYSTEM_KEYWORD_STATE] = function afterDoctypeSystemKeywordState(
         this.state = BEFORE_DOCTYPE_SYSTEM_IDENTIFIER_STATE;
 
     else if (cp === $.QUOTATION_MARK) {
+        this._err(ERR.missingSeparationCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `missing-whitespace-after-doctype-system-keyword`

File: lib/tokenizer/index.js [hunk 14]
```diff
@@ -2110,28 +2137,33 @@ _[AFTER_DOCTYPE_SYSTEM_KEYWORD_STATE] = function afterDoctypeSystemKeywordState(
...
         this.currentToken.systemId = '';
         this.state = DOCTYPE_SYSTEM_IDENTIFIER_DOUBLE_QUOTED_STATE;
     }
 
     else if (cp === $.APOSTROPHE) {
+        this._err(ERR.missingSeparationCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `missing-whitespace-after-doctype-system-keyword`

File: lib/tokenizer/index.js [hunk 15]
```diff
@@ -2110,28 +2137,33 @@ _[AFTER_DOCTYPE_SYSTEM_KEYWORD_STATE] = function afterDoctypeSystemKeywordState(
...
         this.currentToken.systemId = '';
         this.state = DOCTYPE_SYSTEM_IDENTIFIER_SINGLE_QUOTED_STATE;
     }
 
     else if (cp === $.GREATER_THAN_SIGN) {
+        this._err(ERR.unexpectedClosingOfDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `abrupt-closing-of-doctype`

File: lib/tokenizer/index.js [hunk 16]
```diff
@@ -2110,28 +2137,33 @@ _[AFTER_DOCTYPE_SYSTEM_KEYWORD_STATE] = function afterDoctypeSystemKeywordState(
...
 
     else if (cp === $.EOF) {
+        this._err(ERR.eofInDoctype);
         this.currentToken.forceQuirks = true;
         this._emitCurrentToken();
         this._emitEOFToken();
     }
 
     else {
+        this._err(ERR.unexpectedCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `missing-quote-before-doctype-system-identifier`

File: lib/tokenizer/index.js [hunk 17]
```diff
@@ -2155,18 +2187,21 @@ _[BEFORE_DOCTYPE_SYSTEM_IDENTIFIER_STATE] = function beforeDoctypeSystemIdentifi
     }
 
     else if (cp === $.GREATER_THAN_SIGN) {
+        this._err(ERR.unexpectedClosingOfDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `abrupt-closing-of-doctype`

File: lib/tokenizer/index.js [hunk 18]
```diff
@@ -2155,18 +2187,21 @@ _[BEFORE_DOCTYPE_SYSTEM_IDENTIFIER_STATE] = function beforeDoctypeSystemIdentifi
...
 
     else if (cp === $.EOF) {
+        this._err(ERR.eofInDoctype);
         this.currentToken.forceQuirks = true;
         this._emitCurrentToken();
         this._emitEOFToken();
     }
 
     else {
+        this._err(ERR.unexpectedCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `missing-quote-before-doctype-system-identifier`

File: lib/tokenizer/index.js [hunk 19]
```diff
@@ -2179,16 +2214,20 @@ _[DOCTYPE_SYSTEM_IDENTIFIER_DOUBLE_QUOTED_STATE] = function doctypeSystemIdentif
...
         this.state = AFTER_DOCTYPE_SYSTEM_IDENTIFIER_STATE;
 
-    else if (cp === $.NULL)
+    else if (cp === $.NULL) {
+        this._err(ERR.unexpectedNullCharacter);
         this.currentToken.systemId += UNICODE.REPLACEMENT_CHARACTER;
+    }
 
     else if (cp === $.GREATER_THAN_SIGN) {
+        this._err(ERR.unexpectedClosingOfDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `abrupt-closing-of-doctype`

File: lib/tokenizer/index.js [hunk 20]
```diff
@@ -2205,16 +2244,20 @@ _[DOCTYPE_SYSTEM_IDENTIFIER_SINGLE_QUOTED_STATE] = function doctypeSystemIdentif
...
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `abrupt-closing-of-doctype`

File: lib/tokenizer/index.js [hunk 21]
```diff
@@ -2237,13 +2280,16 @@ _[AFTER_DOCTYPE_SYSTEM_IDENTIFIER_STATE] = function afterDoctypeSystemIdentifier
...
     else if (cp === $.EOF) {
+        this._err(ERR.eofInDoctype);
         this.currentToken.forceQuirks = true;
         this._emitCurrentToken();
         this._emitEOFToken();
     }
 
-    else
+    else {
+        this._err(ERR.unexpectedCharacterInDoctype);
```
Comment (inikulin) [inline lib/tokenizer/index.js]: `unexpected-character-after-doctype-system-identifier`

**Review by inikulin — CHANGES_REQUESTED** [review]

**Comment by diervo** [comment]
@inikulin You just keep on not liking my spanglish errors names! xD
Great, I knew you would have good feedback for the names, will change them and push the changes on the rest of repos.

**Comment by inikulin** [comment]
@diervo

> You just keep on not liking my spanglish errors names!

Nah, I just replace them with russinglish variants :D

**Comment by diervo** [comment]
//cc @inikulin had to rebase, but everything ready for the almost last merge.

**Comment by diervo** [comment]
@inikulin Only the last renames are pending, I will try to finish this + tests tomorrow for sure.

**Comment by diervo** [comment]
@inikulin rebased and ready to merge

File: lib/common/error_codes.js [hunk 22]
```diff
@@ -17,14 +17,23 @@ module.exports = {
...
+    unexpectedCharacterInUnquotedAttributeValue: 'unexpected-character-in-unquoted-attribute-value',
     eofBeforeTagName: 'eof-before-tag-name',
     eofInTag: 'eof-in-tag',
     missingAttributeValue: 'missing-attribute-value',
     missingWhitespaceBetweenAttributes: 'missing-whitespace-between-attributes',
-    unexpectedCharacterInUnquotedAttributeValue: 'unexpected-character-in-unquoted-attribute-value',
+    missingWhitespaceAfterDoctypePublicKeyword: 'missing-whitespace-after-doctype-public-keyword',
+    missingWhitespaceBetweenDoctypePublicAndSytemIdentifier: 'missing-whitespace-between-doctype-public-and-system-identifiers',
+    missingWhitespaceAfterDoctypeSystemKeyword: 'missing-whitespace-after-doctype-system-keyword',
+    missingQuoteAroundDoctypePublicIdentifier: 'missing-quote-around-doctype-public-identifier',
```
Comment (inikulin) [inline lib/common/error_codes.js:29]: `missingQuoteAroundDoctypePublicIdentifier` is not used anymore

File: lib/common/error_codes.js [hunk 23]
```diff
@@ -35,7 +44,7 @@ module.exports = {
     nonUnicodeCharacterReference: 'non-unicode-character-reference',
     controlCharacterReference: 'control-character-reference',
     undefinedCharacterReference: 'undefined-character-reference',
-    eofInDoctype: 'eof-in-doctype',
+    controlOrUndefinedCharacterReference: 'control-or-undefined-character-reference',
```
Comment (inikulin) [inline lib/common/error_codes.js:47]: Seems like a rebase artefact. This one is not used anymore.

**Review by inikulin — CHANGES_REQUESTED** [review]

**Comment by inikulin** [comment]
@diervo I guess I'll merge it as is and fix those minor issues. Congratulations! We've done with first huge chunk of work.

**Pull request merged**